src/server.ts

```diff
@@ -31,12 +33,67 @@ const { handleCustomResponse } = require('./custom_response');
 const DEFAULT_AUTH_REALM = 'ProxyChain';
 const DEFAULT_PROXY_SERVER_PORT = 8000;
 
+type ConnectionStats = {
+    srcTxBytes: number;
+    srcRxBytes: number;
+    trgTxBytes: number | null;
+    trgRxBytes: number | null;
+};
+
+type HandlerOpts = {
+    server: Server;
+    id: number;
+    srcRequest: http.IncomingMessage;
+    srcResponse: http.ServerResponse | null;
+    srcHead: Buffer | null;
+    trgParsed: URL | null;
+    upstreamProxyUrlParsed: URL | null;
+    isHttp: boolean;
+    customResponseFunction: CustomResponseOpts['customResponseFunction'] | null;
+};
+
+type PrepareRequestFunctionOpts = {
+    connectionId: unknown;
+    request: http.IncomingMessage;
+    username: string;
+    password: string;
+    hostname: string;
+    port: string;
+    isHttp: boolean;
+};
+
+type PrepareRequestFunctionResult = {
+    customResponseFunction?: CustomResponseOpts['customResponseFunction'];
+    requestAuthentication?: boolean;
+    failMsg?: string;
+    upstreamProxyUrl?: string | null;
+};
+
+type Promisable<T> = T | Promise<T>;
+type PrepareRequestFunction = (opts: PrepareRequestFunctionOpts) => Promisable<undefined | PrepareRequestFunctionResult>;
+
 /**
  * Represents the proxy server.
  * It emits the 'requestFailed' event on unexpected request errors, with the following parameter `{ error, request }`.
  * It emits the 'connectionClosed' event when connection to proxy server is closed, with parameter `{ connectionId, stats }`.
  */
-class Server extends EventEmitter {
+export class Server extends EventEmitter {
+    port: number;
+
+    prepareRequestFunction?: PrepareRequestFunction;
+
+    authRealm: unknown;
+
+    verbose: boolean;
+
+    server: http.Server | null;
```

Comment:
looks like we always have the server set (done in constructor), why the `| null`? there are many `!` later in the code because of it

Reply:
That's because initially it isn't created I think

Reply:
wdym? the instance [gets created in constructor](https://github.com/apify/proxy-chain/pull/161/files/b326729704a75725b2333305289c0ca2e9ce8571#diff-8a8ae07582c9d433ec8c2e5c4310ff8901e604f4965c5b90a49117ad46c47595R145), so it should be there unless you create the instance via `Object.create()`

Reply:
Indeed it is, not sure what I was thinking, I'll fix this

Reply:
Actually it's set to null in the `close` function.

Reply:
I see, still it makes more sense to me to have it defined as mandatory, if you call `close` it feels like you should not use any of the methods anyway.

We are talking about how to type this, the functionality itself shows that we don't check if the property is set, hence it feels like it should be set to mandatory and we could escape the strict type only in the `close` method instead of having `!` almost everywhere.

Reply:
So we can do this in two ways. Should I make it `http.Server` or `const server = this.server!` ?

Reply:
I'd probably rather type it as mandatory instead of introducing local variable with `!` (especially if there are more than one methods that would have to do it). But both are fine.